File: src/main/java/me/meiallu/luck/Lexer.java
package me.meiallu.luck;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Lexer {

    public static List<Token> lex(String path) {
        List<Token> tokenList = new ArrayList<>();

        try {
            BufferedReader bufferedReader = new BufferedReader(new FileReader(path));
            String line = bufferedReader.readLine();

            while (line != null) {
                StringBuilder builder = null;
                char[] chars = line.toCharArray();

                boolean readingString = false;

                for (Character character : chars) {
                    if (character == '"') {
                        if (!readingString) {
                            builder = new StringBuilder();
                            readingString = true;
                        } else {
                            tokenList.add(new Token(TokenType.STRING, builder.toString()));
                            builder = null;
                            readingString = false;
                        }
                        continue;
                    } else if (readingString) {
                        builder.append(character);
                        continue;
                    }

                    switch (character) {
                        case '+', '-', '%', '*':
                            tokenList.add(new Token(TokenType.ARITHMETIC_OPERATOR, character));
                            break;
                        case '!':
                            tokenList.add(new Token(TokenType.LOGICAL_OPERATOR, character));
                            break;
                        case '=':
                            Token lastToken = tokenList.get(tokenList.size() - 1);
                            char lastChar = lastToken.value instanceof Character ? (char) lastToken.value : 'a';

                            if (lastToken.type == TokenType.LOGICAL_OPERATOR && lastChar == '!') {
                                tokenList.remove(lastToken);
                                tokenList.add(new Token(TokenType.COMPARISON_OPERATOR, "!="));
                            } else if (lastToken.type == TokenType.ASSIGNMENT_OPERATOR && lastChar == '=') {
                                tokenList.remove(lastToken);
                                tokenList.add(new Token(TokenType.COMPARISON_OPERATOR, "=="));
                            } else {
                                tokenList.add(new Token(TokenType.ASSIGNMENT_OPERATOR, character));
                            }
                            break;
                        case '<', '>':
                            tokenList.add(new Token(TokenType.COMPARISON_OPERATOR, character));
                            break;
                        case ' ', '(', ')', ':':
                            if (builder != null) {
                                tokenList.add(getToken(builder));
                                builder = null;
                            }

                            if (character == ':')
                                tokenList.add(new Token(TokenType.BLOCK_START, character));

                            break;
                        default:
                            if (builder == null)
                                builder = new StringBuilder();

                            builder.append(character);
                            break;
                    }
                }

                if (builder != null)
                    tokenList.add(getToken(builder));

                if (tokenList.get(tokenList.size() - 1).type != TokenType.NEW_LINE)
                    tokenList.add(new Token(TokenType.NEW_LINE, null));

                line = bufferedReader.readLine();
            }

            bufferedReader.close();
        } catch (IOException exception) {
            throw new RuntimeException(exception);
        }

        return tokenList;
    }

    public static Token getToken(StringBuilder builder) {
        String toString = builder.toString();

        switch (toString) {
            case "&&", "||":
                return new Token(TokenType.LOGICAL_OPERATOR, toString);
            case "fun", "while", "if", "elif", "else":
                return new Token(TokenType.KEYWORD, toString);
            case "byte", "short", "int", "long", "float", "double", "char", "boolean":
                return new Token(TokenType.TYPE, toString);
            case "true", "false":
                return new Token(TokenType.BOOLEAN, toString);
            default:
                try {
                    double parsedDouble = Double.parseDouble(toString);
                    return new Token(TokenType.NUMBER, parsedDouble);
                } catch (NumberFormatException ignored) {
                    return new Token(TokenType.IDENTIFIER, toString);
                }
        }
    }
}

File: src/main/kotlin/runner/Lexer.kt
package me.meiallu.luck.runner

import me.meiallu.luck.data.Token
import me.meiallu.luck.data.TokenType
import java.io.File

class Lexer {

    companion object {
        fun lex(path: String): ArrayList<Token> {
            val tokenList: ArrayList<Token> = ArrayList()

            File(path).forEachLine {
                var builder: StringBuilder? = null
                val chars = it.toCharArray()
                var readingString = false

                for (character in chars) {

                    if (character == '"') {
                        if (!readingString) {
                            builder = StringBuilder()
                            readingString = true
                        } else {
                            tokenList += Token(TokenType.STRING, builder.toString())
                            builder = null
                            readingString = false
                        }
                        continue
                    } else if (readingString) {
                        builder?.append(character)
                        continue
                    }

                    when (character) {
                        '+', '-', '*', '/', '%' -> tokenList += Token(TokenType.ARITHMETIC_OPERATOR, character)
                        '=' -> tokenList += Token(TokenType.ASSIGNMENT_OPERATOR, null)
                        '{' -> tokenList += Token(TokenType.BLOCK_START, null)
                        '}' -> tokenList += Token(TokenType.BLOCK_END, null)
                        '<', '>' -> tokenList += Token(TokenType.COMPARISON_OPERATOR, character)
                        ' ', ',', '(', ')' -> {
                            if (builder != null) {
                                val token = getToken(builder)
                                tokenList += token

                                if (character == '(' && token.type == TokenType.IDENTIFIER)
                                    token.type = TokenType.FUNCTION

                                builder = null
                            }

                            when (character) {
                                ')' -> tokenList += Token(TokenType.FUNCTION_END, null)
                                ',' -> tokenList += Token(TokenType.SEPARATOR, null)
                            }
                        }
                        else -> {
                            if (builder == null)
                                builder = StringBuilder()

                            builder.append(character)
                        }
                    }
                }

                if (builder != null)
                    tokenList += getToken(builder)

                if (tokenList[tokenList.size - 1].type != TokenType.NEW_LINE)
                    tokenList += Token(TokenType.NEW_LINE, null)
            }
            return tokenList
        }

        private fun getToken(builder: java.lang.StringBuilder): Token {
            val toString = builder.toString()

            when (toString) {
                "var", "const" -> return Token(TokenType.TYPE, toString)
                "if" -> return Token(TokenType.KEYWORD, toString)
            }

            return if (toString.toDoubleOrNull() != null)
                Token(TokenType.NUMBER, toString.toDouble())
            else
                Token(TokenType.IDENTIFIER, toString)
        }
    }
}